sklearn/metrics/_classification.py

```diff
@@ -321,17 +321,25 @@ def confusion_matrix(y_true, y_pred, *, labels=None, sample_weight=None,
                          "'all', None}")
 
     n_labels = labels.size
-    label_to_ind = {y: x for x, y in enumerate(labels)}
-    # convert yt, yp into index
-    y_pred = np.array([label_to_ind.get(x, n_labels + 1) for x in y_pred])
-    y_true = np.array([label_to_ind.get(x, n_labels + 1) for x in y_true])
+    # If labels are not consecutive integers starting from zero, then
+    # yt and yp must be converted into index form
+    need_index_conversion = not (
+        labels.dtype.kind in {'i', 'u', 'b'} and
+        np.all(labels == np.arange(n_labels)) and
+        y_true.min() >= 0 and y_pred.min() >= 0
+    )
+    if need_index_conversion:
+        label_to_ind = {y: x for x, y in enumerate(labels)}
+        y_pred = np.array([label_to_ind.get(x, n_labels + 1) for x in y_pred])
+        y_true = np.array([label_to_ind.get(x, n_labels + 1) for x in y_true])
 
     # intersect y_pred, y_true with labels, eliminate items not in labels
     ind = np.logical_and(y_pred < n_labels, y_true < n_labels)
-    y_pred = y_pred[ind]
-    y_true = y_true[ind]
-    # also eliminate weights of eliminated items
-    sample_weight = sample_weight[ind]
+    if not np.all(ind):
```

Comment:
Does including this greatly improve performance?

I think we can revert this change and only have `needs_index_conversion` so we can get this merged quicker.

Reply:
I remember it being significant when I originally wrote this, which was awhile ago. It will depend on the use case, but I believe it does give a noticeable speed up to the fastest case with a minimal impact on the slow cases.

I would have to benchmark again to be sure.

Reply:
So I was able to come up with a small benchmark that at least demonstrates the efficacy of the idea. In most cases the true and pred labels are all going to be "valid" -- i.e. between 0 and n_labels. In this case the question is: do we gain a benefit by checking if we _should_ do the `__getitem__` operation before we do it? Or is the `__getitem__` fast enough such that the check isn't saving us time. 

Furthermore, we have the less common case where there are invalid labels. In that case, does doing this extra check (which by definition will fail) cost us that much in overhead?

To test this I came up with this code, which simply times the `all` and the `__getitem__` operation.

```python
    import ubelt as ub
    import numpy as np
    results = []
    ns = np.logspace(1, 7, 100).astype(np.int)
    for n in ub.ProgIter(ns, desc='time-tradeoff', verbose=3):
        print('n = {!r}'.format(n))
        y_true = np.random.randint(0, 100, n).astype(np.int64)
        y_pred = np.random.randint(0, 100, n).astype(np.int64)
        sample_weight = np.random.rand(n)

        isvalid = np.random.rand(n) > 0

        import timerit
        ti = timerit.Timerit(9, bestof=3, verbose=2)
        for timer in ti.reset('all-check'):
            with timer:
                np.all(isvalid)
        results.append({
            'n': n,
            'label': ti.label,
            'time': ti.mean(),
        })

        for timer in ti.reset('all-index'):
            with timer:
                y_true[isvalid]
                y_pred[isvalid]
                sample_weight[isvalid]
        results.append({
            'n': n,
            'label': ti.label,
            'time': ti.mean(),
        })

    df = pd.DataFrame(results)

    import kwplot
    import seaborn as sns
    kwplot.autoplt()
    sns.set()
    ax = sns.lineplot(data=df, x='n', y='time', hue='label')
    ax.set_yscale('log')
    ax.set_xscale('log')
```

which produces 

![image](https://user-images.githubusercontent.com/3186211/110735789-733a3880-81f8-11eb-9270-d396db3129ca.png)

So in the case where there are more than 1000 elements (the case that we start to really care about speed!), we get an order of magnitude difference in how long each of those lines takes.

In the case where we must incur both costs, again the check is an order of magnitude faster than the actual `__getitem__` call, so it doesn't add much overhead.

Reply:
Now the question is, how much does it matter in the grand scheme of this function?

Well, running the benchmark suite we see on average this line takes about 0.8% of the total time (which in this case was 0.735286 s)

```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   270                                               # intersect y_pred, y_true with labels, eliminate items not in labels
   271       324       4191.0     12.9      0.6      ind = np.logical_and(y_pred < n_labels, y_true < n_labels)
   272       324       5657.0     17.5      0.8      if not np.all(ind):
   273                                                   y_pred = y_pred[ind]
   274                                                   y_true = y_true[ind]
   275                                                   # also eliminate weights of eliminated items
   276                                                   sample_weight = sample_weight[ind]
```

Now if we comment that portion of the code and reprofile we see:

```
   270                                               # intersect y_pred, y_true with labels, eliminate items not in labels
   271       324       4272.0     13.2      0.6      ind = np.logical_and(y_pred < n_labels, y_true < n_labels)
   272                                               # if not np.all(ind):
   273       324       2541.0      7.8      0.4      y_pred = y_pred[ind]
   274       324       2061.0      6.4      0.3      y_true = y_true[ind]
   275                                               # also eliminate weights of eliminated items
   276       324       2182.0      6.7      0.3      sample_weight = sample_weight[ind]
```

which takes about 1.0% of the time on average. So there is benefit to having this check! But not much. However, that's only part of the story. The benchmark script is running over inputs of size 10, 1000, and 10000, and we know from above that this starts to make a difference when the input is larger. So how much?

Well with the check we see:

```
   270                                               # intersect y_pred, y_true with labels, eliminate items not in labels
   271       216      17120.0     79.3      0.4      ind = np.logical_and(y_pred < n_labels, y_true < n_labels)
   272       216       5704.0     26.4      0.1      if not np.all(ind):
   273                                                   y_pred = y_pred[ind]
   274                                                   y_true = y_true[ind]
   275                                                   # also eliminate weights of eliminated items
   276                                                   sample_weight = sample_weight[ind]
```

and without we get

```
   270                                               # intersect y_pred, y_true with labels, eliminate items not in labels
   271       216      17222.0     79.7      0.4      ind = np.logical_and(y_pred < n_labels, y_true < n_labels)
   272                                               # if not np.all(ind):
   273       216      19397.0     89.8      0.4      y_pred = y_pred[ind]
   274       216      18590.0     86.1      0.4      y_true = y_true[ind]
   275                                               # also eliminate weights of eliminated items
   276       216      22554.0    104.4      0.5      sample_weight = sample_weight[ind]
```

so that's a 0.1% of the time versus 1.3% of the time. That is a very significant difference, and I believe is enough to justify existence of the line. Let me know what you think. @thomasjpfan

Reply:
Honestly we don't worry about 1% of a function call (if that percent stays constant with the dataset size). But has this specific code snippet is simple, I am fine with keeping it.